video/out/wayland_common.c

```diff
@@ -1552,10 +1564,39 @@ static void apply_keepaspect(struct vo_wayland_state *wl, int *width, int *heigh
     if (!wl->opts->keepaspect)
         return;
 
+    int phys_width = handle_round(wl->scaling, *width);
+    int phys_height = handle_round(wl->scaling, *height);
+
+    // Ensure that the size actually changes before we start trying to actually
+    // calculate anything so the wrong direction isn't choosen.
+    if (wl->resizing && !wl->resizing_direction &&
+        phys_width == mp_rect_w(wl->geometry) && phys_height == mp_rect_h(wl->geometry))
+        return;
+
+    // We are doing a continuous resize (e.g. dragging with mouse), keep the
+    // aspect ratio against the height if the change is only in the height
+    // coordinate.
+    if (wl->resizing && !wl->resizing_direction && phys_width == mp_rect_w(wl->geometry) &&
+        phys_height != mp_rect_h(wl->geometry)) {
```

Comment:
> Well, it seems pretty unintuitive that when I drag a corner of the window, the resizing direction depends on the movement direction of the first pixel moved, which can be very inconsistent with human hand movements.

What do you mean? The resizing direction is is reset for every configure event, what compositor do you see this behavior on?

Reply:
> Unfortunately there seems to be no way in Wayland to determine which edge/corner is being used for resize if the resizing comes from server-side decoration.

We don't really need this information? We have enough information to deduce which dimension (if any) we're constrained in while resizing, that's all the information we require for the resize. It's upto the compositor to "place" the window correctly, which sway doesn't because it's buggy, but river handles mpv resizes perfectly. As does Weston from my testing.

Reply:
I called it "resizing_direction" because I'm bad at naming but in case it's not clear it just means whether you constrain based on the width or the height.

Reply:
I think just adding a comment should suffice

Reply:
It ends when the user lifts the mouse button, however, you receive a new configure event with the `resizing` state whenever the compositor wants us to resize or the state changes. So we receive a configure with `(960, 540, resizing | activated)` -> `(960, 541, resizing | activated)`, and so on.

Reply:
If you initially go vertically and then proceed to move the mouse perfectly horizontally afterwards, sure it wouldn't resize if that's what you mean.

Reply:
Keep in mind that the current behavior of resizing from a corner if you go straight up is "it doesn't work at all". I don't see what the practical problem is here. If your first pixel move is straight up, it is extremely likely you are moving the mouse in a mostly vertical direction anyway. I'll concede that if you do something weird like start in one direction and then perfectly move the mouse in another direction, it won't resize but that seems really contrived and unnatural.